cpu/cortexm_common/cortexm_init.c

```diff
@@ -72,3 +72,40 @@ void cortexm_init(void)
     cortexm_init_isr_priorities();
     cortexm_init_misc();
 }
+
+static const uint32_t BFARVALID_MASK = (0x80 << SCB_CFSR_BUSFAULTSR_Pos);
+
+bool cpu_check_address(volatile const char *address)
+{
+#if defined(CPU_ARCH_CORTEX_M3) || defined(CPU_ARCH_CORTEX_M4) || \
+    defined(CPU_ARCH_CORTEX_M4F) || defined(CPU_ARCH_CORTEX_M7)
+    bool is_valid = true;
+
+    /* Clear BFARVALID flag */
+    SCB->CFSR |= BFARVALID_MASK;
+
+    /* Ignore BusFault by enabling BFHFNMIGN */
+    SCB->CCR |= SCB_CCR_BFHFNMIGN_Msk;
+    __asm volatile ("cpsid f;");
+
+    *address;
+    /* Check BFARVALID flag */
+    if ((SCB->CFSR & BFARVALID_MASK) != 0)
+    {
+        /* Bus Fault occured reading the address */
+        is_valid = false;
+    }
+
+    __asm volatile ("cpsie f;");
```

Comment:
What if the mask was disabled when calling this function. They will be enabled. Maybe it's better to save/restore the flag.

Reply:
I'm pretty sure people should not play with FAULTMASK without a very solid reason, so it should be enabled at all times except when doind some very specific things.

Also, not sure if it's a good idea to write FAULTMASK register directly - there's MSR/MRS instructions to do so, but I can't find if there are any undesired consequences compared to using CPS instructions. I've tried and MSR/MRS (__get_FAULTMASK / __set_FAULTMASK from CMSIS) work, but still not sure if it's ok to use it so.

Regular interrupts (controlled by irq_enable / irq_disable) won't change anyway, it's PRIMASK, not FAULTMASK.

P.S. Changed assember to inline CMSIS instructions.

Reply:
> I'm pretty sure people should not play with FAULTMASK without a very solid reason, so it should be enabled at all times except when doind some very specific things.

Agreed, but I wanted to prevent any surprising weird interaction.

> Also, not sure if it's a good idea to write FAULTMASK register directly

If you look at https://github.com/RIOT-OS/RIOT/blob/master/cpu/cortexm_common/irq_arch.c it is doing that (it's even using cmsis builtins). I'm not sure why it is doing what it is doing in `irq_enable` (I just opened a bug report -  #10076).

It may work to have a fault_disable/fault_restore.

Reply:
Ok, as no one sees no harm either, I've changed CPS to direct access with MRS with FAULTMASK state preservation. Moved BFHFNMIGN after interrupts disables too.

P.S. Don't think we need separate fault_disable function in public API.